src/tasks/components/TaskScheduler/ScriptSelector.tsx

```diff
@@ -1,44 +1,149 @@
 
-export const ScriptSelector: FC = () => {
+interface Props {
+  selectedScript: Script
+  setSelectedScript: (script: Script) => void
+}
+
+export const ScriptSelector: FC<Props> = ({
+  selectedScript,
+  setSelectedScript,
+}) => {
+  const [scripts, setScripts] = useState<Script[]>([])
+  const [searchTerm, setSearchTerm] = useState('')
+  const [scriptsLoadingStatus, setScriptsLoadingStatus] = useState(
+    RemoteDataState.NotStarted
+  )
+  const dispatch = useDispatch()
   const {orgID} = useParams<{orgID: string}>()
   const history = useHistory()
+
   const fluxScriptEditorLink = `/orgs/${orgID}/data-explorer?fluxScriptEditor`
 
+  useEffect(() => {
+    setScriptsLoadingStatus(RemoteDataState.Loading)
+    const getScripts = async () => {
+      try {
+        const scripts = await fetchScripts()
+        setScripts(scripts.scripts)
+        setScriptsLoadingStatus(RemoteDataState.Done)
+      } catch (error) {
+        setScriptsLoadingStatus(RemoteDataState.Error)
+        dispatch(notify(getScriptsFail()))
+      }
+    }
+    getScripts()
+  }, [dispatch])
+
+  const filteredScripts = () =>
+    scripts.filter(script =>
+      script.name.toLocaleLowerCase().includes(searchTerm.toLocaleLowerCase())
+    )
+
+  const searchForTerm = (event: ChangeEvent<HTMLInputElement>) => {
+    setSearchTerm(event.target.value)
+  }
+
   const openScriptEditor = () => {
     history.push(`/orgs/${orgID}/data-explorer?fluxScriptEditor`)
   }
 
-  const createScriptInEditor = (
-    <Dropdown.Item onClick={openScriptEditor}>
-      + Create Script in Editor
-    </Dropdown.Item>
-  )
+  const handleSelectScript = script => {
+    setSelectedScript(script)
+  }
+
+  let scriptsList
 
-  const dropdownButtonText = 'Select a Script'
+  if (
+    scriptsLoadingStatus === RemoteDataState.NotStarted ||
+    scriptsLoadingStatus === RemoteDataState.Loading
+  ) {
+    scriptsList = (
+      <div>
+        <TechnoSpinner strokeWidth={ComponentSize.Small} diameterPixels={32} />
+      </div>
+    )
+  }
+
+  if (scriptsLoadingStatus === RemoteDataState.Error) {
+    scriptsList = (
+      <div>
+        <p>Could not get scripts</p>
+      </div>
+    )
+  }
+
+  if (scriptsLoadingStatus === RemoteDataState.Done) {
+    scriptsList = (
+      <>
+        {filteredScripts().map(script => (
+          <Dropdown.Item
+            key={script.id}
+            value={script.name}
+            onClick={() => handleSelectScript(script)}
+            selected={script.name === selectedScript?.name}
+          >
+            {script.name}
+          </Dropdown.Item>
+        ))}
+      </>
+    )
+    if (!filteredScripts().length && searchTerm) {
+      scriptsList = (
+        <EmptyState>
+          <p>{`No Scripts match "${searchTerm}"`}</p>
+        </EmptyState>
+      )
+    } else if (!filteredScripts().length && !searchTerm) {
+      scriptsList = (
+        <EmptyState>
+          <p>No Scripts found</p>
+        </EmptyState>
+      )
+    }
+  }
```

Comment:
This does filtering up to 3 times. Once to set scriptsList, once in the if condition, and possibly once more in the if-else condition.

We should just filter once. Something like this:
```
  if (scriptsLoadingStatus === RemoteDataState.Done) {
    const filteredScripts = filterScripts()
    scriptsList = (
      <>
        {filteredScripts.map(script => (
          <Dropdown.Item
            key={script.id}
            value={script.name}
            onClick={() => handleSelectScript(script)}
            selected={script.name === selectedScript?.name}
          >
            {script.name}
          </Dropdown.Item>
        ))}
      </>
    )
    if (!filteredScripts.length) {
      if (searchTerm) {
        scriptsList = (
          <EmptyState>
            <p>{`No Scripts match "${searchTerm}"`}</p>
          </EmptyState>
        )
      } else {
        scriptsList = (
          <EmptyState>
            <p>No Scripts found</p>
          </EmptyState>
        )
      }
    }
  }
```

Be sure to note the name change for the function in the other comments.

Reply:
great point. I'll make the change